torch_xla/csrc/xla_lower_util.cpp

```diff
@@ -694,6 +694,24 @@ xla::XlaOp CreatePut(const Device& device, xla::XlaOp input, xla::XlaOp index,
   return XlaHelpers::DynamicReshapeAs(r1_scatter, input_shape);
 }
 
+xla::XlaOp CreateLinspace(const Device& device, xla::XlaOp start,
+                          xla::XlaOp end, int64_t steps) {
+  std::tie(start, end) = XlaHelpers::PromoteValues(start, end);
+  if (steps == 1) {
+    return BuildExpand(start, {1});
```

Comment:
question, if steps == 1, should we still promote the `start`? We should check pytorch behavior for this.

Reply:
when we refer to dtpye, we usually mean `at::ScalarType`. There are currently two type for a XLATensor.

`at::ScalarType` is the type we inherit/set from the pytorch frontend. It represents how pytorch think/want the type of certain tensor should be.

`xla::PrimitveType` are the actual type we store in the XLA device side. If user didn't set `at::ScalarType` explictly, we will convert the `xla::PrimitveType` to the corresponding `at::scalartype` if pytorch frontend ask for it.


For the most part these type should match, but for example if user do
```
a = torch.tensor(int32)
b = torch.tensor(int32)
c = torch.add(a, b, dtype = int64)
```

what we ended up doing is `xla::PrimitiveType` of `c` will still be `int32`(inferred by xla compiler) but the `ScalarType` are `int64`. We do this to make pytorch frontend happy.


What I want to say here is that `dtype` should not affect the xla compiler backend lowering logic. In this case we want to make the result of the `linspace`'s `xla::Primitivetype` respect the `start` and `end`. Instead of being promoitved to int64 due to index is of type int64.

Reply:
I converted `start` and `end` to F32 when creating the `ir::Value`s in `tensor_methods.cpp`, since the PyTorch docs say `start` and `end` are taken as `float`s. If I kept the type of start and end as `int` and the float division result of `(end - start) / steps` was not the same as the int division result, I got a different output compared to `torch.linspace` without XLA. So PyTorch definitely promotes the types to floats internally.

 I mainly added the promotion as a sanity check to avoid a crash in case we change something at a higher level or another function calls `CreateLinspace`. It will be a no-op if you're actually calling `torch.linspace`. Since we're passing in start and end as floats, IIUC they can not be promoted to int64; index and steps will instead be promoted from int to float.